java/integTest/src/test/java/glide/PubSubTests.java

```diff
@@ -0,0 +1,893 @@
+        var listenerSharded =
+                createListener(
+                        false, useCallback, PubSubClusterChannelMode.SHARDED.ordinal(), subscriptionsSharded);
+
+        var sender = (RedisClusterClient) createClient(false);
+        clients.addAll(List.of(listenerExact, listenerPattern, listenerSharded, sender));
+
+        for (var pubsubMessage : exactMessages) {
+            sender.publish(pubsubMessage.getChannel(), pubsubMessage.getMessage()).get();
+        }
+        for (var pubsubMessage : patternMessages) {
+            sender.publish(pubsubMessage.getChannel(), pubsubMessage.getMessage()).get();
+        }
+        for (var pubsubMessage : shardedMessages) {
+            sender.spublish(pubsubMessage.getChannel(), pubsubMessage.getMessage()).get();
+        }
+
+        Thread.sleep(MESSAGE_DELIVERY_DELAY); // deliver the messages
+
+        if (useCallback) {
+            var expected = new HashSet<Pair<Integer, PubSubMessage>>();
+            expected.addAll(
+                    exactMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.EXACT.ordinal(), m))
+                            .collect(Collectors.toSet()));
+            expected.addAll(
+                    patternMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.PATTERN.ordinal(), m))
+                            .collect(Collectors.toSet()));
+            expected.addAll(
+                    shardedMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.SHARDED.ordinal(), m))
+                            .collect(Collectors.toSet()));
+
+            verifyReceivedPubsubMessages(expected, listenerExact, useCallback);
+        } else {
+            verifyReceivedPubsubMessages(
+                    exactMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.EXACT.ordinal(), m))
+                            .collect(Collectors.toSet()),
+                    listenerExact,
+                    useCallback);
+            verifyReceivedPubsubMessages(
+                    patternMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.PATTERN.ordinal(), m))
+                            .collect(Collectors.toSet()),
+                    listenerPattern,
+                    useCallback);
+            verifyReceivedPubsubMessages(
+                    shardedMessages.stream()
+                            .map(m -> Pair.of(PubSubClusterChannelMode.SHARDED.ordinal(), m))
+                            .collect(Collectors.toSet()),
+                    listenerSharded,
+                    useCallback);
+        }
+    }
+
+    /**
+     * Similar to `test_pubsub_three_publishing_clients_same_name_with_sharded` in python client
+     * tests.
+     */
+    @SneakyThrows
+    @Test
+    public void three_publishing_clients_same_name_with_sharded_no_callback() {
```

Comment:
why do we have no_callback flavor ?

Reply:
I split parametrized test to 2 non paramtrized, due to big differences

Reply:
align with python

Reply:
We will never achieve perfect alignment between languages due to language differences.  Let's focus on getting all the features over the line.

Reply:
I've made three_publishing_clients_same_name_with_sharded() parameterized on whether callbacks are on now in #1773 .